File: src/main/java/com/starshootercity/abilities/PhantomizeOverlay.java
package com.starshootercity.abilities;

import com.destroystokyo.paper.event.player.PlayerPostRespawnEvent;
import net.kyori.adventure.key.Key;
import net.minecraft.network.protocol.game.ClientboundSetBorderWarningDistancePacket;
import net.minecraft.world.level.border.WorldBorder;
import org.bukkit.craftbukkit.v1_20_R3.entity.CraftPlayer;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;
import org.jetbrains.annotations.NotNull;

public class PhantomizeOverlay implements DependantAbility, Listener {
    @Override
    public @NotNull Key getKey() {
        return Key.key("origins:phantomize_overlay");
    }

    @Override
    public @NotNull Key getDependencyKey() {
        return Key.key("origins:phantomize");
    }

    @EventHandler
    public void onPhantomizeToggle(Phantomize.PhantomizeToggleEvent event) {
        updatePhantomizeOverlay(event.getPlayer());
    }

    @EventHandler
    public void onPlayerPostRespawn(PlayerPostRespawnEvent event) {
        updatePhantomizeOverlay(event.getPlayer());
    }

    @EventHandler
    public void onPlayerJoin(PlayerJoinEvent event) {
        updatePhantomizeOverlay(event.getPlayer());
    }

    private void updatePhantomizeOverlay(Player player) {
        WorldBorder border = new WorldBorder();
        border.setWarningBlocks(getDependency().isEnabled(player) ? (int) player.getWorld().getWorldBorder().getSize() * 2 : player.getWorld().getWorldBorder().getWarningDistance());
        ClientboundSetBorderWarningDistancePacket warningDistancePacket = new ClientboundSetBorderWarningDistancePacket(border);
        ((CraftPlayer) player).getHandle().connection.send(warningDistancePacket);
    }
}

File: core/src/main/kotlin/ru/turbovadim/abilities/PhantomizeOverlay.kt
package ru.turbovadim.abilities

import com.destroystokyo.paper.event.player.PlayerPostRespawnEvent
import ru.turbovadim.OriginsRebornEnhanced.Companion.NMSInvoker
import ru.turbovadim.abilities.Phantomize.PhantomizeToggleEvent
import net.kyori.adventure.key.Key
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerJoinEvent

class PhantomizeOverlay : DependantAbility, Listener {

    override fun getKey(): Key {
        return Key.key("origins:phantomize_overlay")
    }

    override val dependencyKey: Key = Key.key("origins:phantomize")

    @EventHandler
    fun onPhantomizeToggle(event: PhantomizeToggleEvent) {
        updatePhantomizeOverlay(event.getPlayer())
    }

    @EventHandler
    fun onPlayerPostRespawn(event: PlayerPostRespawnEvent) {
        updatePhantomizeOverlay(event.getPlayer())
    }

    @EventHandler
    fun onPlayerJoin(event: PlayerJoinEvent) {
        updatePhantomizeOverlay(event.getPlayer())
    }

    private fun updatePhantomizeOverlay(player: Player) {
        NMSInvoker.setWorldBorderOverlay(player, dependency.isEnabled(player))
    }
}
